packages/wallet/src/Wallet.d.ts

```diff
@@ -0,0 +1,308 @@
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint &
+    InternalCaveatSpecification = InternalCaveatSpecification,
+> =
+  | AccountsControllerActions
+  | ApprovalControllerActions
+  | CurrencyRateControllerActions
+  | GasFeeControllerActions
+  | KeyringControllerActions
+  | NetworkControllerActions
+  | PermissionControllerActions
+  | PreferencesControllerActions
+  | TransactionControllerActions
+  | WalletActions<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >;
+
+export type WalletStateChange<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint &
+    InternalPermissionSpecification = InternalPermissionSpecification,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint &
+    InternalCaveatSpecification = InternalCaveatSpecification,
+> = ControllerStateChangeEvent<
+  'Wallet',
+  MetamaskState<
+    ControllerPermissionSpecification,
+    ControllerCaveatSpecification
+  >
+>;
+
+export type WalletEvents<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint &
+    InternalPermissionSpecification = InternalPermissionSpecification,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint &
+    InternalCaveatSpecification = InternalCaveatSpecification,
+> = WalletStateChange<
+  ControllerPermissionSpecification,
+  ControllerCaveatSpecification
+>;
+
+/**
+ * All wallet events.
+ */
+export type AllWalletEvents<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint &
+    InternalPermissionSpecification = InternalPermissionSpecification,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint &
+    InternalCaveatSpecification = InternalCaveatSpecification,
+> =
+  | AccountsControllerEvents
+  | ApprovalControllerEvents
+  | CurrencyRateControllerEvents
+  | GasFeeControllerEvents
+  | KeyringControllerEvents
+  | NetworkControllerEvents
+  | PermissionControllerEvents
+  | PreferencesControllerEvents
+  | TransactionControllerEvents
+  | WalletEvents<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >;
+
+/**
+ * A MetaMask wallet.
+ *
+ * @template ControllerPermissionSpecification - A union of the types of all
+ * permission specifications available to the controller. Any referenced caveats
+ * must be included in the controller's caveat specifications.
+ * @template ControllerCaveatSpecification - A union of the types of all
+ * caveat specifications available to the controller.
+ */
+export class MetamaskWallet<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint &
+    InternalPermissionSpecification = InternalPermissionSpecification,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint &
+    InternalCaveatSpecification = InternalCaveatSpecification,
+> {
+  #controllerMessenger: ControllerMessenger<AllWalletActions, AllWalletEvents>;
+
+  #controllers: {
```

Comment:
Here is the initial set of controllers that I'll be proposing for the MVP.

I wanted enough functionality to demonstrate that the JSON-RPC pipeline works and is useful, so the NetworkController was a must. I brought in the TransactionController as well because I wanted at least one intercepted method, and it seemed easier to bring that in than the SignatureController (because it's already on BaseControllerV2, but the message managers are not).

And the GasFeeController was brought in because it's a peer dependency of the TransactionController. Same for the ApprovalController. Though in both cases there are other reasons to include them as well: the GasFeeController is a nice example of a polling controller, and the ApprovalController establishes how we handle user confirmations.

CurrencyRateController was the one asset controller that seemed easy to bring on its own, and it lets us demonstrate more complex wallet API interactions down the line (e.g. when we go to write an integration test to approximate what the UI might want to render a confirmation).

AccountsController and PreferencesController were brought in for the identity state, which is still in both places at the moment. Even after it's removed from preferences, wallet preferences is nice to have on hand. Later I'll be suggesting that we audit the preferences to ensure they're all relevant to the wallet library, moving anything else to a MobilePreferencesController.

Lastly, KeyringController and PermissionController need to be included, they're fundamental to what the wallet is.

Reply:
I was tempted to include snaps even in this early draft, but that seemed pretty complicated to integrate. And it's not completely landed on mobile yet.

Plus, snaps provides a nice demonstration of why we need extensible permissions.